### PR TITLE
Adds support for batch events

### PR DESCRIPTION
**Pull request by zimbatm**

### CODE REVIEW

**Comment by jpatel531** [comment]
@zimbatm - would it be possible to remove the duplication between the `Event` struct and `eventPayload` struct. maybe through embedding common fields?

File: client.go [hunk 1]
```diff
@@ -190,6 +190,26 @@ func (c *Client) trigger(channels []string, event string, data interface{}, sock
 	return unmarshalledBufferedEvents(response)
 }
 
+type BatchRequest struct {
```
Comment (Nicowcow) [inline client.go]: Is it worth making the `BatchRequest` public? None of the functions allow it as a parameter

File: client.go [hunk 2]
```diff
@@ -190,6 +190,26 @@ func (c *Client) trigger(channels []string, event string, data interface{}, sock
...
+	Batch []Event `json:"batch"`
+}
+
+func (c *Client) TriggerBatch(batch []Event) (*BufferedEvents, error) {
```
Comment (Nicowcow) [inline client.go]: I guess the reason for using a new `Event` type here instead of having something similar to `Trigger` is to avoid receiving arrays of different lengths, and I guess the reason `Trigger` is not using an `Event` is for backwards compatibility... yet the inconsistency is a bit confusing. Also, the event name is called `event` in `Trigger` and `Name` in the `Event` type. Would it make sense to add a new function that has the same behavior as `Trigger` but is called with an `Event` instead? Then we could rename that new function to `Trigger` in the next major release.

File: client.go [hunk 3]
```diff
@@ -190,6 +190,26 @@ func (c *Client) trigger(channels []string, event string, data interface{}, sock
...
+	payload, err := json.Marshal(&BatchRequest{batch})
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("/apps/%s/batch_events", c.AppId)
+	u := createRequestURL("POST", c.Host, path, c.Key, c.Secret, authTimestamp(), c.Secure, payload, nil, c.Cluster)
```
Comment (Nicowcow) [inline client.go]: Last minor comment (very personal): there 6 similar `createRequestURL` calls, which (in this file) always takes a `Client`, and always calls `authTimestamp` to get the timestamp. Might be worth creating a helper, which would also drop the dependency on `time`.

Comment (jpatel531) [inline client.go]: yeah the reason i introduced the time dependency was so it would be easier to unit test. open to new methods thought

File: client.go (hunk 1 above)
Comment (zimbatm) [inline client.go]: Fixed in 58d5b147ab41f6b8620e011faa30d4fb349f7f44

File: client.go (hunk 2 above)
Comment (zimbatm) [inline client.go]: In the case of `event` vs `Name` I think it's more `event` vs `event.Name` so I would rename the first one to `eventName`.

I'm not sure if we can compact all the trigger methods. `trigger(Event[]{{"event", "channel", "data"}})` doesn't seem much of an improvement, and it doesn't support the multi-channel scenario.

File: client.go (hunk 3 above)
Comment (zimbatm) [inline client.go]: +1 for refactoring the library. it probably belongs to another PR though.

**Comment by zimbatm** [comment]
@jpatel531 in regard to Event vs eventPayload, I started doing that but then realized the eventPayload is not exactly the same since it takes a list of channels and we don't support that in the batch trigger.

File: client.go (hunk 2 above)
Comment (zimbatm) [inline client.go]: Fixed the eventName consistency in 76ad82ae6d9e504dd816b4c6dcc3818a3c20d418

**Comment by zimbatm** [comment]
@jpatel531 @Nicowcow do I get an ACK ?

**Comment by jpatel531** [comment]
I'd ideally like to validate channels and socket ids before triggering, but I'm not going to exist on it for now

File: client.go (hunk 3 above)
Comment (jpatel531) [inline client.go]: yeah - i had only known go for like 3 days before writing the library. i'd probably do it a lot cleaner if i did it now
